From 0d9a293b8a06bb7ed393e27f32e93f94f79b427a Mon Sep 17 00:00:00 2001
From: Amithash Prasad <amithash@meta.com>
Date: Thu, 1 May 2025 15:34:59 -0700
Subject: [PATCH 21/21] platform-mc: Add structured logging for threshold
 events

Create structured threshold log events up creation
and alert triggers. These will also clear previously
raised logs.

Tested:
Bump the fan speed to low and let the CPU heat up.
Ensure we emit logs on Upper Critical and Upper Hard Shutdown.
```
root@sled325945102-oob:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/19 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData property  a{ss}  8 "READING_VALUE" "90.125" "SENSOR_NAME" "/xyz/openbmc_project/sensors/temperature/Sentinel_Dome_Slot_1_MB_CPU_TEMP_C" "THRESHOLD_VALUE" "90.0" "UNITS" "xyz.openbmc_project.Sensor.Value.Unit.DegreesC" <snip>
.Message property  s "xyz.openbmc_project.Sensor.Threshold.ReadingAboveUpperCriticalThreshold" emits-change writable
.Resolved property  b false emits-change writable

root@sled325945102-oob:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/21 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData property  a{ss} 8 "READING_VALUE" "95.125" "SENSOR_NAME" "/xyz/openbmc_project/sensors/temperature/Sentinel_Dome_Slot_1_MB_CPU_TEMP_C" "THRESHOLD_VALUE" "95.0" "UNITS" "xyz.openbmc_project.Sensor.Value.Unit.DegreesC" <snip>
.Message property  s "xyz.openbmc_project.Sensor.Threshold.ReadingAboveUpperHardShutdownThreshold" emits-change writable
.Resolved property  b         false emits-change writable
```
Bump the fan speed back to high and let the CPU cool down.
Ensure we emit the SensorReadingNormalRange and that the
existing logs are marked as resolved.
```
root@sled325945102-oob:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/22 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData property  a{ss} 7 "READING_VALUE" "85.75" "SENSOR_NAME" "/xyz/openbmc_project/sensors/temperature/Sentinel_Dome_Slot_1_MB_CPU_TEMP_C" "UNITS" "xyz.openbmc_project.Sensor.Value.Unit.DegreesC" <snip>
.Message property  s  "xyz.openbmc_project.Sensor.Threshold.SensorReadingNormalRange" change writable
.Resolved property  b false change writable

root@sled325945102-oob:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/19 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData property  a{ss} 8 "READING_VALUE" "90.125" "SENSOR_NAME" "/xyz/openbmc_project/sensors/temperature/Sentinel_Dome_Slot_1_MB_CPU_TEMP_C" "THRESHOLD_VALUE" "90.0" "UNITS" "xyz.openbmc_project.Sensor.Value.Unit.DegreesC" <snip>
.Message property  s  "xyz.openbmc_project.Sensor.Threshold.ReadingAboveUpperCriticalThreshold" emits-change writable
.Resolved property  b true emits-change writable

root@sled325945102-oob:~# busctl introspect -l xyz.openbmc_project.Logging /xyz/openbmc_project/logging/entry/21 xyz.openbmc_project.Logging.Entry | grep "AdditionalData\|Message\|Resolved"
.AdditionalData property  a{ss}     8 "READING_VALUE" "95.125" "SENSOR_NAME" "/xyz/openbmc_project/sensors/temperature/Sentinel_Dome_Slot_1_MB_CPU_TEMP_C" "THRESHOLD_VALUE" "95.0" "UNITS" "xyz.openbmc_project.Sensor.Value.Unit.DegreesC" <snip>
.Message property  s "xyz.openbmc_project.Sensor.Threshold.ReadingAboveUpperHardShutdownThreshold" emits-change writable
.Resolved property  b true emits-change writable
```

Change-Id: I72623f7f929a4d7b17f87d102d34747ff6cab3ae
Signed-off-by: Amithash Prasad <amithash@meta.com>
---
 platform-mc/numeric_sensor.cpp          | 149 ++++++++++++++++++++++++
 platform-mc/numeric_sensor.hpp          |  14 +++
 platform-mc/test/event_manager_test.cpp |  12 +-
 3 files changed, 174 insertions(+), 1 deletion(-)

diff --git a/platform-mc/numeric_sensor.cpp b/platform-mc/numeric_sensor.cpp
index 7536fe4..88b1ecf 100644
--- a/platform-mc/numeric_sensor.cpp
+++ b/platform-mc/numeric_sensor.cpp
@@ -5,6 +5,11 @@
 
 #include <libpldm/platform.h>
 
+#include <phosphor-logging/commit.hpp>
+#include <sdbusplus/asio/property.hpp>
+#include <xyz/openbmc_project/Logging/Entry/client.hpp>
+#include <xyz/openbmc_project/Sensor/Threshold/event.hpp>
+
 #include <limits>
 #include <regex>
 
@@ -14,6 +19,13 @@ namespace pldm
 {
 namespace platform_mc
 {
+
+// Allow unit tests to disable threshold logging to ensure
+// we don't bring in additional dependencies.
+bool PLDM_THRESHOLD_LOGGING = true;
+
+// This allows code to cleanly iterate through all supported
+// threshold levels and directions.
 static const std::array<pldm::utils::Level, 3> allThresholdLevels = {
     pldm::utils::Level::WARNING, pldm::utils::Level::CRITICAL,
     pldm::utils::Level::HARDSHUTDOWN};
@@ -785,6 +797,20 @@ bool NumericSensor::checkThreshold(bool alarm, bool direction, double value,
     }
     return alarm;
 }
+
+bool NumericSensor::hasThresholdAlarm()
+{
+    bool alarm = false;
+    for (auto level : allThresholdLevels)
+    {
+        for (auto direction : allThresholdDirections)
+        {
+            alarm |= getThresholdAlarm(level, direction);
+        }
+    }
+    return alarm;
+}
+
 void NumericSensor::setWarningThresholdAlarm(pldm::utils::Direction direction,
                                              double value, bool newAlarm)
 {
@@ -902,9 +928,132 @@ int NumericSensor::setThresholdAlarm(pldm::utils::Level level,
         default:
             return PLDM_ERROR;
     }
+    if (newAlarm)
+    {
+        createThresholdLog(level, direction, value);
+    }
+    else
+    {
+        auto& log = assertedLog[{level, direction}];
+        if (log.has_value())
+        {
+            clearThresholdLog(log);
+        }
+        // If all alarms have cleared. Log normal range.
+        if (!hasThresholdAlarm())
+        {
+            createNormalRangeLog(value);
+        }
+    }
     return PLDM_SUCCESS;
 }
 
+void NumericSensor::clearThresholdLog(
+    std::optional<sdbusplus::message::object_path>& log)
+{
+    if (PLDM_THRESHOLD_LOGGING)
+    {
+        if (log)
+        {
+            try
+            {
+                lg2::resolve(*log);
+            }
+            catch (std::exception& ec)
+            {
+                lg2::error("Error trying to resolve: {LOG} : {ERROR}", "LOG",
+                           std::string(*log), "ERROR", ec.what());
+            }
+            log.reset();
+        }
+    }
+}
+
+template <typename errorObj>
+auto logThresholdHelper(const std::string& sensorObjPath, double value,
+                        SensorUnit sensorUnit, double threshold)
+    -> std::optional<sdbusplus::message::object_path>
+{
+    if (PLDM_THRESHOLD_LOGGING)
+    {
+        try
+        {
+            return lg2::commit(
+                errorObj("SENSOR_NAME", sensorObjPath, "READING_VALUE", value,
+                         "UNITS", sensorUnit, "THRESHOLD_VALUE", threshold));
+        }
+        catch (std::exception& ec)
+        {
+            lg2::error("Unable to create log entry for {SENSOR} : {ERROR}",
+                       "SENSOR", sensorObjPath, "ERROR", ec.what());
+        }
+    }
+    return std::nullopt;
+}
+
+void NumericSensor::createThresholdLog(
+    pldm::utils::Level level, pldm::utils::Direction direction, double value)
+{
+    namespace Errors =
+        sdbusplus::error::xyz::openbmc_project::sensor::Threshold;
+    std::string sensorObjPath = sensorNameSpace + sensorName;
+    double threshold = getThreshold(level, direction);
+    static const std::map<
+        std::tuple<pldm::utils::Level, pldm::utils::Direction>,
+        std::function<std::optional<sdbusplus::message::object_path>(
+            const std::string&, double, SensorUnit, double)>>
+        thresholdEventMap = {
+            {{pldm::utils::Level::WARNING, pldm::utils::Direction::HIGH},
+             &logThresholdHelper<Errors::ReadingAboveUpperWarningThreshold>},
+            {{pldm::utils::Level::WARNING, pldm::utils::Direction::LOW},
+             &logThresholdHelper<Errors::ReadingBelowLowerWarningThreshold>},
+            {{pldm::utils::Level::CRITICAL, pldm::utils::Direction::HIGH},
+             &logThresholdHelper<Errors::ReadingAboveUpperCriticalThreshold>},
+            {{pldm::utils::Level::CRITICAL, pldm::utils::Direction::LOW},
+             &logThresholdHelper<Errors::ReadingBelowLowerCriticalThreshold>},
+            {{pldm::utils::Level::HARDSHUTDOWN, pldm::utils::Direction::HIGH},
+             &logThresholdHelper<
+                 Errors::ReadingAboveUpperHardShutdownThreshold>},
+            {{pldm::utils::Level::HARDSHUTDOWN, pldm::utils::Direction::LOW},
+             &logThresholdHelper<
+                 Errors::ReadingBelowLowerHardShutdownThreshold>},
+        };
+    try
+    {
+        auto log = thresholdEventMap.at(
+            {level, direction})(sensorObjPath, value, sensorUnit, threshold);
+        assertedLog[{level, direction}] = log;
+    }
+    catch (std::exception& ec)
+    {
+        lg2::error(
+            "Unable to create threshold log entry for {OBJPATH}: {ERROR}",
+            "OBJPATH", sensorObjPath, "ERROR", ec.what());
+    }
+}
+
+void NumericSensor::createNormalRangeLog(double value)
+{
+    if (PLDM_THRESHOLD_LOGGING)
+    {
+        namespace Events =
+            sdbusplus::event::xyz::openbmc_project::sensor::Threshold;
+        std::string objPath = sensorNameSpace + sensorName;
+        try
+        {
+            lg2::commit(Events::SensorReadingNormalRange(
+                "SENSOR_NAME", objPath, "READING_VALUE", value, "UNITS",
+                sensorUnit));
+        }
+        catch (std::exception& ec)
+        {
+            lg2::error(
+                "Unable to create SensorReadingNormalRange log entry for {OBJPATH}: {ERROR}",
+                "OBJPATH", objPath, "ERROR", ec.what());
+        }
+    }
+}
+
 void NumericSensor::updateThresholds()
 {
     double value = std::numeric_limits<double>::quiet_NaN();
diff --git a/platform-mc/numeric_sensor.hpp b/platform-mc/numeric_sensor.hpp
index b09bcd4..430374d 100644
--- a/platform-mc/numeric_sensor.hpp
+++ b/platform-mc/numeric_sensor.hpp
@@ -294,6 +294,12 @@ class NumericSensor
         return false;
     }
 
+    /* @brief Returns true if at least one threshold alarm is set
+     *
+     * @return true if at least one threshold alarm is set
+     */
+    bool hasThresholdAlarm();
+
     /* @brief raises the alarm on the warning threshold
      *
      * @param[in] direction - The threshold direction (HIGH/LOW)
@@ -380,6 +386,10 @@ class NumericSensor
     SensorUnit sensorUnit;
 
   private:
+    void clearThresholdLog(std::optional<sdbusplus::message::object_path>& log);
+    void createNormalRangeLog(double value);
+    void createThresholdLog(pldm::utils::Level level,
+                            pldm::utils::Direction direction, double value);
     /**
      * @brief Check sensor reading if any threshold has been crossed and update
      * Threshold interfaces accordingly
@@ -431,6 +441,10 @@ class NumericSensor
     /** @brief A power-of-10 multiplier for baseUnit */
     int8_t baseUnitModifier;
     bool useMetricInterface = false;
+
+    std::map<std::tuple<pldm::utils::Level, pldm::utils::Direction>,
+             std::optional<sdbusplus::message::object_path>>
+        assertedLog;
 };
 } // namespace platform_mc
 } // namespace pldm
diff --git a/platform-mc/test/event_manager_test.cpp b/platform-mc/test/event_manager_test.cpp
index 0b64b0c..5c46738 100644
--- a/platform-mc/test/event_manager_test.cpp
+++ b/platform-mc/test/event_manager_test.cpp
@@ -16,6 +16,14 @@
 using ::testing::_;
 using ::testing::Return;
 
+namespace pldm
+{
+namespace platform_mc
+{
+extern bool PLDM_THRESHOLD_LOGGING;
+}
+} // namespace pldm
+
 class EventManagerTest : public testing::Test
 {
   protected:
@@ -27,7 +35,9 @@ class EventManagerTest : public testing::Test
         terminusManager(event, reqHandler, instanceIdDb, termini, nullptr),
         eventManager(terminusManager, termini),
         platformManager(terminusManager, termini, nullptr)
-    {}
+    {
+        pldm::platform_mc::PLDM_THRESHOLD_LOGGING = false;
+    }
 
     PldmTransport* pldmTransport = nullptr;
     sdbusplus::bus_t& bus;
-- 
2.47.1

